cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1145 @@
+	dbInfo, ok := s.doSchemaByID(id)
+	if !ok {
+		return cerror.ErrSnapshotSchemaNotFound.GenWithStackByArgs(id)
+	}
+	tag := negative(currentTs)
+	s.schemas.ReplaceOrInsert(newVersionedID(id, tag))
+	s.schemaNameToID.ReplaceOrInsert(newVersionedEntityName(-1, dbInfo.Name.O, tag))
+	for _, id := range s.tablesInSchema(dbInfo.Name.O) {
+		tbInfo, _ := s.doPhysicalTableByID(id)
+		s.doDropTable(tbInfo, currentTs)
+	}
+	s.currentTs = currentTs
+	log.Debug("drop schema success", zap.String("name", dbInfo.Name.O), zap.Int64("id", dbInfo.ID))
+	return nil
+}
+
+// Create a new schema in the snapshot. `dbInfo` will be deep copied.
+func (s *Snapshot) createSchema(dbInfo *timodel.DBInfo, currentTs uint64) error {
+	x, ok := s.doSchemaByID(dbInfo.ID)
+	if ok {
+		return cerror.ErrSnapshotSchemaExists.GenWithStackByArgs(x.Name, x.ID)
+	}
+	if id, ok := s.doSchemaIDByName(dbInfo.Name.O); ok {
+		return cerror.ErrSnapshotSchemaExists.GenWithStackByArgs(dbInfo.Name.O, id)
+	}
+	s.doCreateSchema(dbInfo, currentTs)
+	s.currentTs = currentTs
+	log.Debug("create schema success", zap.String("name", dbInfo.Name.O), zap.Int64("id", dbInfo.ID))
+	return nil
+}
+
+// Replace a schema. dbInfo will be deep copied.
+// Callers should ensure `dbInfo` information not conflict with other schemas.
+func (s *Snapshot) replaceSchema(dbInfo *timodel.DBInfo, currentTs uint64) error {
+	old, ok := s.doSchemaByID(dbInfo.ID)
+	if !ok {
+		return cerror.ErrSnapshotSchemaNotFound.GenWithStack("schema %s(%d) not found", dbInfo.Name, dbInfo.ID)
+	}
+	s.doCreateSchema(dbInfo, currentTs)
+	if old.Name.O != dbInfo.Name.O {
+		tag := negative(currentTs)
+		s.schemaNameToID.ReplaceOrInsert(newVersionedEntityName(-1, old.Name.O, tag))
+	}
+	s.currentTs = currentTs
+	log.Debug("replace schema success", zap.String("name", dbInfo.Name.O), zap.Int64("id", dbInfo.ID))
+	return nil
+}
+
+func (s *Snapshot) doCreateSchema(dbInfo *timodel.DBInfo, currentTs uint64) {
+	tag := negative(currentTs)
+	vid := newVersionedID(dbInfo.ID, tag)
+	vid.target = unsafe.Pointer(dbInfo.Clone())
+	s.schemas.ReplaceOrInsert(vid)
+	vname := newVersionedEntityName(-1, dbInfo.Name.O, tag)
+	vname.target = dbInfo.ID
+	s.schemaNameToID.ReplaceOrInsert(vname)
+}
+
+// dropTable removes a table(NOT partition) from the snapshot.
+func (s *Snapshot) dropTable(id int64, currentTs uint64) error {
+	tbInfo, ok := s.doPhysicalTableByID(id)
+	if !ok {
+		return cerror.ErrSnapshotTableNotFound.GenWithStackByArgs(id)
+	}
+	s.doDropTable(tbInfo, currentTs)
+	s.currentTs = currentTs
+	log.Debug("drop table success",
+		zap.String("schema", tbInfo.TableName.Schema),
+		zap.String("table", tbInfo.TableName.Table),
+		zap.Int64("id", tbInfo.ID))
+	return nil
+}
+
+func (s *Snapshot) doDropTable(tbInfo *model.TableInfo, currentTs uint64) {
+	tag := negative(currentTs)
+	s.tables.ReplaceOrInsert(newVersionedID(tbInfo.ID, tag))
+	s.tableNameToID.ReplaceOrInsert(newVersionedEntityName(tbInfo.SchemaID, tbInfo.TableName.Table, tag))
```

Comment:
The name `Snapshot` suggests it's read-only, but this method mutates `Snapshot`, should we change the name?

Reply:
I'm not sure so I prefer to keep less changes in this PR.